plugins/cairo-lang-quote/src/lib.rs

```diff
@@ -0,0 +1,103 @@
+use std::iter::Peekable;
+
+use proc_macro::{Delimiter, TokenStream as RustTokenStream, TokenTree as RustTokenTree};
+use proc_macro2::{Ident, Span};
+
+extern crate proc_macro;
+use quote::quote as rust_quote;
+
+#[derive(Debug)]
+enum QuoteToken {
+    Var(Ident),
+    Content(String),
+}
+
+impl QuoteToken {
+    pub fn from_delimiter(delimiter: Delimiter, is_end: bool) -> Self {
+        match (delimiter, is_end) {
+            (Delimiter::Brace, false) => Self::Content("{".to_string()),
+            (Delimiter::Brace, true) => Self::Content("}".to_string()),
+            (Delimiter::Bracket, false) => Self::Content("[".to_string()),
+            (Delimiter::Bracket, true) => Self::Content("]".to_string()),
+            (Delimiter::Parenthesis, false) => Self::Content("(".to_string()),
+            (Delimiter::Parenthesis, true) => Self::Content(")".to_string()),
+            (Delimiter::None, _) => Self::Content(String::default()),
+        }
+    }
+}
+
+fn process_token_stream(
+    mut token_stream: Peekable<impl Iterator<Item = RustTokenTree>>,
+    output: &mut Vec<QuoteToken>,
+) {
+    // Rust proc macro parser to TokenStream gets rid of all whitespaces.
+    // Here we just make sure no two identifiers are without a space between them.
+    let mut was_last_ident: bool = false;
+    while let Some(token_tree) = token_stream.next() {
+        match token_tree {
+            RustTokenTree::Group(group) => {
+                let token_iter = group.stream().into_iter().peekable();
+                let delimiter = group.delimiter();
+                output.push(QuoteToken::from_delimiter(delimiter, false));
+                process_token_stream(token_iter, output);
+                output.push(QuoteToken::from_delimiter(delimiter, true));
+                was_last_ident = false;
+            }
+            RustTokenTree::Punct(punct) => {
+                if punct.as_char() == '#' {
+                    if let Some(RustTokenTree::Ident(ident)) = token_stream.next() {
+                        let var_ident = Ident::new(&ident.to_string(), Span::call_site());
+                        output.push(QuoteToken::Var(var_ident))
+                    }
+                } else {
+                    output.push(QuoteToken::Content(punct.to_string()));
+                }
+                was_last_ident = false;
+            }
+            RustTokenTree::Ident(ident) => {
+                if was_last_ident {
+                    output.push(QuoteToken::Content(format!(" {}", ident)));
```

Comment:
Maybe it would be better to add `QuoteToken::Whitespace` and move whitespace logic to the place where you create a token stream out of it?